diff --git a/node_modules/react-native-nitro-modules/android/src/main/cpp/utils/JNISharedPtr.hpp b/node_modules/react-native-nitro-modules/android/src/main/cpp/utils/JNISharedPtr.hpp
index 41bc012..e7afe1c 100644
--- a/node_modules/react-native-nitro-modules/android/src/main/cpp/utils/JNISharedPtr.hpp
+++ b/node_modules/react-native-nitro-modules/android/src/main/cpp/utils/JNISharedPtr.hpp
@@ -22,7 +22,7 @@ struct GlobalRefDeleter {
 
   void operator()(T* /* cthis */) {
     // It's RAII - once `GlobalRefDeleter` goes out of scope, `jni::global_ref` will too.
-    _ref = nullptr;
+    jni::ThreadScope::WithClassLoader([&] { _ref = nullptr; });
   }
 
 private:
diff --git a/node_modules/react-native-nitro-modules/cpp/core/HybridFunction.hpp b/node_modules/react-native-nitro-modules/cpp/core/HybridFunction.hpp
index aefd987..c2e06fb 100644
--- a/node_modules/react-native-nitro-modules/cpp/core/HybridFunction.hpp
+++ b/node_modules/react-native-nitro-modules/cpp/core/HybridFunction.hpp
@@ -23,6 +23,10 @@ struct JSIConverter;
 #include <string>
 #include <type_traits>
 
+#ifdef ANDROID
+#include <fbjni/fbjni.h>
+#endif
+
 namespace margelo::nitro {
 
 using namespace facebook;
@@ -118,6 +122,10 @@ public:
         std::string funcName = getHybridFuncFullName<THybrid>(kind, name, hybridInstance.get());
         std::string message = exception.what();
         throw jsi::JSError(runtime, funcName + ": " + message);
+      } catch (const jni::JniException& exception) {
+        std::string funcName = getHybridFuncFullName<THybrid>(kind, name, hybridInstance.get());
+        std::string message = exception.what();
+        throw jsi::JSError(runtime, funcName + ": " + message);
 #pragma clang diagnostic pop
 #endif
       } catch (...) {
